sklearn/model_selection/_search.py

```diff
@@ -513,84 +518,139 @@ def _fit(self, X, y, labels, parameter_iterable):
             n_candidates = len(parameter_iterable)
             print("Fitting {0} folds for each of {1} candidates, totalling"
                   " {2} fits".format(n_splits, n_candidates,
                                      n_candidates * n_splits))
+        candidate_params = list(parameter_iterable)
+        # Get all the unique parameters across all the candidates
+        parameter_names = sorted(set(chain(*map(six.iterkeys,
+                                                candidate_params))))
 
         base_estimator = clone(self.estimator)
-
         pre_dispatch = self.pre_dispatch
 
         out = Parallel(
             n_jobs=self.n_jobs, verbose=self.verbose,
             pre_dispatch=pre_dispatch
         )(delayed(_fit_and_score)(clone(base_estimator), X, y, self.scorer_,
                                   train, test, self.verbose, parameters,
-                                  self.fit_params, return_parameters=True,
+                                  self.fit_params, return_parameters=False,
                                   error_score=self.error_score)
-          for parameters in parameter_iterable
+          for parameters in candidate_params
           for train, test in cv.split(X, y, labels))
 
-        # Out is a list of triplet: score, estimator, n_test_samples
-        n_fits = len(out)
-
-        scores = list()
-        grid_scores = list()
-        for grid_start in range(0, n_fits, n_splits):
-            n_test_samples = 0
-            score = 0
-            all_scores = []
-            for this_score, this_n_test_samples, _, parameters in \
-                    out[grid_start:grid_start + n_splits]:
-                all_scores.append(this_score)
-                if self.iid:
-                    this_score *= this_n_test_samples
-                    n_test_samples += this_n_test_samples
-                score += this_score
-            if self.iid:
-                score /= float(n_test_samples)
-            else:
-                score /= float(n_splits)
-            scores.append((score, parameters))
-            # TODO: shall we also store the test_fold_sizes?
-            grid_scores.append(_CVScoreTuple(
-                parameters,
-                score,
-                np.array(all_scores)))
-        # Store the computed scores
-        self.grid_scores_ = grid_scores
-
-        # Find the best parameters by comparing on the mean validation score:
-        # note that `sorted` is deterministic in the way it breaks ties
-        best = sorted(grid_scores, key=lambda x: x.mean_validation_score,
-                      reverse=True)[0]
-        self.best_params_ = best.parameters
-        self.best_score_ = best.mean_validation_score
+        n_candidates = len(candidate_params)
+
+        test_scores, test_sample_counts, _ = zip(*out)
+
+        test_scores = np.array(
+            test_scores, dtype=np.float64).reshape(n_candidates, n_splits)
```

Comment:
FWIW, we tend to prefer the other style of permitted line wrapping where possible:

```python
test_scores = np.array(test_scores,
                       dtype=np.float64).reshape(n_candidates, n_splits)
```